## Proyecto de aula
## Entrega 2: Ingenieria de caracteristicas

En este notebook se desarrollara la entrega 2 de la materia de analisis de datos de ingenieria de sistemas.

# Carga y comprension de datos




`a`. Carguen el archivo de datos en pandas

In [15]:
import pandas as pd
from pathlib import Path 
import matplotlib.pyplot as plt
import seaborn as sns

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"

columns = [
    "age", "sex", "cp", "trestbps", "chol", "fbs", "restecg",
    "thalach", "exang", "oldpeak", "slope", "ca", "thal", "target"
]


ROOT_DIR = Path().resolve().parent

DATA_DIR =  ROOT_DIR / "data/raw"

file_path = 'processed.cleveland.data'

df = pd.read_csv(
  
  url, names = columns,
  #na_values = ["?", "NA", "NaN", "nan", ""]
  
  )

df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


`b`. Indique cúantas variables y registros tiene el dataframe, y asegúrense de que el tipo de dato de cada variable sea el esperado.


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        303 non-null    str    
 12  thal      303 non-null    str    
 13  target    303 non-null    int64  
dtypes: float64(11), int64(1), str(2)
memory usage: 33.3 KB


`conclusion`: Se puede apreciar que en el df hay 14 columnas, y 303 filas.

`conclusion`: Se puede apreciar como las variables 'ca' y 'thal' tienen como tipo str, lo que es incorreco ya que son datos float64.
Debido a esto, se revisa y encuentra que estas variables presentan varios registros nulos, por lo que se definen como nulos con el objetivo de que pandas escoja correctamente el tipado.

In [4]:
df = pd.read_csv(
  
  url, names = columns,
  na_values = ["?", "NA", "NaN", "nan", ""]
  
  )

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    float64
 1   sex       303 non-null    float64
 2   cp        303 non-null    float64
 3   trestbps  303 non-null    float64
 4   chol      303 non-null    float64
 5   fbs       303 non-null    float64
 6   restecg   303 non-null    float64
 7   thalach   303 non-null    float64
 8   exang     303 non-null    float64
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    float64
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(13), int64(1)
memory usage: 33.3 KB


<p style='color: red'>`c`. Reporte las estadisticas descriptivas de las variables numericas y categoricas </p>

<p style='color: white'> Antes de continuar, necesitamos convertir algunas variables que aparentan ser numericas, a si version categorica para su respectivo analisis. Estas puede verlas en la imagen en el archivo importations.ipynb </p>

In [7]:
# Conversion de variables
df['sex'] = df['sex'].astype(str)
df['cp'] = df['cp'].astype(str)
df['fbs'] = df['fbs'].astype(str)
df['restecg'] = df['restecg'].astype(str)
df['exang'] = df['exang'].astype(str)
df['slope'] = df['slope'].astype(str)
df['thal'] = df['thal'].astype(str)


<p style='color: red'>`Numericas` </p>

In [9]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,297.0,54.542088,9.049736,29.0,48.0,56.0,61.0,77.0
trestbps,297.0,131.693603,17.762806,94.0,120.0,130.0,140.0,200.0
chol,297.0,247.350168,51.997583,126.0,211.0,243.0,276.0,564.0
thalach,297.0,149.599327,22.941562,71.0,133.0,153.0,166.0,202.0
oldpeak,297.0,1.055556,1.166123,0.0,0.0,0.8,1.6,6.2
ca,297.0,0.676768,0.938965,0.0,0.0,0.0,1.0,3.0
target,297.0,0.946128,1.234551,0.0,0.0,0.0,2.0,4.0


<h2> Analisis Exploratorio de variables Categoricas </h2>

<p style='color: red'> `estadisticas` </p> 

In [10]:
df.describe(include='object').T

/tmp/ipykernel_129158/1760094569.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object').T


,count,unique,top,freq
sex,297,2,1.0,201
cp,297,4,4.0,142
fbs,297,2,0.0,254
restecg,297,3,0.0,147
exang,297,2,0.0,200
slope,297,3,1.0,139
thal,297,3,3.0,164


<p style='color: red'> Cardinalidad </p> 

In [ ]:
for cat in df.select_dtypes(include=['object']).columns:
    plt.figure()
    sns.countplot(data=df, y=cat, order=df[cat].value_counts().index)
    plt.show();

sex : <StringArray>
['1.0', '0.0']
Length: 2, dtype: str
cp : <StringArray>
['1.0', '4.0', '3.0', '2.0']
Length: 4, dtype: str
fbs : <StringArray>
['1.0', '0.0']
Length: 2, dtype: str
restecg : <StringArray>
['2.0', '0.0', '1.0']
Length: 3, dtype: str
exang : <StringArray>
['0.0', '1.0']
Length: 2, dtype: str
slope : <StringArray>
['3.0', '2.0', '1.0']
Length: 3, dtype: str
thal : <StringArray>
['6.0', '3.0', '7.0']
Length: 3, dtype: str


/tmp/ipykernel_129158/34278814.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  var_categoricas = df.select_dtypes(include="object").columns


`d`. Identifique si el dataframe tiene datos duplicados

In [31]:
df.duplicated().sum()

np.int64(0)

`conclusion`: Se puede apreciar como el dataframe no tiene valores duplicados.

`e`. Identifique si hay variables con valores nulos. 

In [32]:
df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

In [6]:
df = df.dropna(subset=['ca', 'thal'])
df.info()

<class 'pandas.DataFrame'>
Index: 297 entries, 0 to 301
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       297 non-null    float64
 1   sex       297 non-null    float64
 2   cp        297 non-null    float64
 3   trestbps  297 non-null    float64
 4   chol      297 non-null    float64
 5   fbs       297 non-null    float64
 6   restecg   297 non-null    float64
 7   thalach   297 non-null    float64
 8   exang     297 non-null    float64
 9   oldpeak   297 non-null    float64
 10  slope     297 non-null    float64
 11  ca        297 non-null    float64
 12  thal      297 non-null    float64
 13  target    297 non-null    int64  
dtypes: float64(13), int64(1)
memory usage: 34.8 KB


`conclusion`: Se puede apreciar como las variables 'ca' y 'thal' presentan valores nulos, lo que causo aque al inicio se les asignara erroneamente el tipo de dato. como solo hablamos de un total de 6 registros con datos nulos, correspondientes al 2% del dataset, eliminamos estos datos.

`f`. Reporten hallazgos en una celda de texto (Desarrollado a lo largo del taller) 

`g`. Descarten las variables que consideren irrelevantes para su modelo.

Necesitamos realizar el estudio de los datos para saber cuales pueden ser descartadas

In [ ]:
df.head()

,sex,cp,trestbps,chol,restecg,thalach,exang,oldpeak,slope,target
0,1.0,1.0,145.0,233.0,2.0,150.0,0.0,2.3,3.0,0
1,1.0,4.0,160.0,286.0,2.0,108.0,1.0,1.5,2.0,2
2,1.0,4.0,120.0,229.0,2.0,129.0,1.0,2.6,2.0,1
3,1.0,3.0,130.0,250.0,0.0,187.0,0.0,3.5,3.0,0
4,0.0,2.0,130.0,204.0,2.0,172.0,0.0,1.4,1.0,0


`Conclusion`: Se descartaron las variables 'ca', 'thal y fbs, ya que las dos primeras presentan muchos datos nulos, por lo que se descarta, y se descarta fbs debido a que tiene poca varianza y mas del 85% es 0

`h`: Gestionen los datos duplicados y los datos nulos.

In [34]:
df = df.drop_duplicates()
df.isnull().sum()

sex         0
cp          0
trestbps    0
chol        0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
target      0
dtype: int64

`Conclusion`: Ya que anteriormente se descartaron las unicas columnas con datos nulos, no se debe descartar ninguna columna en este paso. Ademas, se descartan los registros duplicados.

# 2. Analisis exploratorio de variables categóricas

`Conclusion`: Debido a que el dataframe no posee variables categoricas, el analisis exploratorio de variables categóricas no se puede realizar.

# 3	Análisis exploratorio de variables numéricas:

`a`. Identifiquen variables que sean (aproximadamente) normales y otras que tengan una distribución diferente.